src/Utilities/StdArrayHelpers.hpp

```diff
@@ -166,6 +175,64 @@ decltype(auto) magnitude(const std::array<T, 3>& a) noexcept {
 }
 //@}
 
+/// \ingroup UtilitiesGroup
+/// \brief Dot product between two arrays
+///
+/// \details This also works elementwise if T is a container and R is a float or
+/// the other way round. The return type will always be the iterable type. If
+/// both or none are iterable, the return type will be T, the first argument
+/// type.
+///
+/// \requires The return type must have operator+= implemented. Elementwise
+/// multiplication between T and R must be defined. In addition, the sizes of
+/// all containers within T and R must match. If T and R are both iterable, the
+/// size of each container in R must be greater or equal than the size of each
+/// container in T.
+
+// `DotReturnType_t` will always be T, unless R is iterable and T is not
+// iterable.
+template <typename T, typename R, typename = std::nullptr_t>
+struct DotReturnType {
+  using type = tt::remove_cvref_wrap_t<T>;
+};
+
+template <typename T, typename R>
+struct DotReturnType<
+    T, R, Requires<not tt::is_iterable_v<T> and tt::is_iterable_v<R>>> {
+  using type = tt::remove_cvref_wrap_t<R>;
+};
+
+template <typename T, typename R>
+using DotReturnType_t = typename DotReturnType<T, R>::type;
+
+template <size_t Dim, typename T, typename R>
+void dot(const gsl::not_null<tt::remove_cvref_wrap_t<T>*> res,
```

Comment:
I think either take the return type as another template parameter, or use `DotReturnType` here as well. Currently I think this function isn't symmetric in `first` and `second`.